### PR TITLE
Performance: speed up room list filtering

### PR DESCRIPTION
**Pull request by bwindels**
https://github.com/vector-im/riot-web/issues/8058

### CODE REVIEW

**Comment by ara4n** [comment]
hum. comparing this branch on my account versus develop, with the test being the time the app locks up when hitting delete on an filter-that-doesn't-match-anything (i.e. go from fully filtered to no filter):

screenshot 2019-01-31 at 17 53 01
screenshot 2019-01-31 at 17 52 21

i.e. 1.5s (this branch) v 1.8s (develop).  I'm not sure this is as big a win as we might desire...

File: src/components/views/rooms/RoomList.js [hunk 1]
```diff
@@ -574,12 +574,29 @@ module.exports = React.createClass({
...
+        const filteredFlags = list.map((room) => {
+            const matches = (
+                    room.name &&
+                    room.name.toLowerCase().includes(lcFilter)
+                ) || (
+                    filter[0] === '#' &&
+                    room.getAliases().some((alias) => {
+                        return alias.toLowerCase().startsWith(lcFilter);
+                    })
+                );
```
Comment (turt2live) [inline src/components/views/rooms/RoomList.js]: This variable is hard to read: I'd recommend breaking it out into multiple other variables and doing an OR on those.

(it was also hard to read before, and could probably use some additional variables as well)

File: src/components/views/rooms/RoomList.js [hunk 2]
```diff
@@ -614,23 +631,37 @@ module.exports = React.createClass({
...
-            return len !== 0 || props.onAddRoom;
-        }));
+        function isSubListVisible(props, count = props.list.length) {
+            return visibleCount(props, count) !== 0 || props.onAddRoom;
+        }
+
+        subListsProps = subListsProps.filter((props) => isSubListVisible(props));
+
+        subListsProps.forEach((props) => {
+            const {count, filteredFlags} = this._applySearchFilter(props.list, this.props.searchFilter);
```
Comment (turt2live) [inline src/components/views/rooms/RoomList.js]: `_mapSubListProps` is called on every `render`, which then calls the relatively CPU intensive `_applySearchFilter`, making this the likely source of Matthew's problems. Caching the result of `_applySearchFilter` (or better yet: pre-calculating it ahead of a render) would yield better results from the `render` at least, and probably make the experience nicer for users.

Comment (turt2live) [reply src/components/views/rooms/RoomList.js]: alternatively, cache the whole `_mapSubListProps` to avoid all the hits

**Comment by bwindels** [comment]
Abandoning PR, in favour of yet to be implemented virtual rendering (just rendering the RoomTiles visible within the scrollview)

**Pull request closed without merging**